src/devices/cerOdometryDevice/CerOdometryDevice.cpp

```diff
@@ -0,0 +1,229 @@
+        yCWarning(CERODOMDEVICE) << "no local name found, using nothing";
+    }  else {
+        localName = config.find("local").asString();
+    }
+
+    //reset odometry
+    resetOdometry();
+
+    //get robot geometry
+    yarp::os::Bottle geometry_group = config.findGroup("ROBOT_GEOMETRY");
+    if (geometry_group.isNull())
+    {
+        yCError(CERODOMDEVICE, "CerOdometryDevice::open Unable to find ROBOT_GEOMETRY group!");
+        return false;
+    }
+    if (!geometry_group.check("geom_r"))
+    {
+        yCError(CERODOMDEVICE, "Missing param geom_r in [ROBOT_GEOMETRY] group");
+        return false;
+    }
+    if (!geometry_group.check("geom_L"))
+    {
+        yCError(CERODOMDEVICE, "Missing param geom_L in [ROBOT_GEOMETRY] group");
+        return false;
+    }
+    geom_r = geometry_group.find("geom_r").asFloat64();
+    geom_L = geometry_group.find("geom_L").asFloat64();
+
+    return true;
+}
+
+void CerOdometryDevice::compute()
+{
+    std::lock_guard lock(m_odometry_mutex);
+    if (ienc) {
+        //read the encoders (deg)
+        ienc->getEncoder(0, &encL);
+        ienc->getEncoder(1, &encR);
+
+        //read the speeds (deg/s)
+        ienc->getEncoderSpeed(0, &velL);
+        ienc->getEncoderSpeed(1, &velR);
+
+        //remove the offset and convert in radians
+        enc[0] = (encL - encL_offset) * 0.0174532925;
+        enc[1] = (encR - encR_offset) * 0.0174532925;
+
+        //estimate the speeds
+        iCub::ctrl::AWPolyElement el;
+        el.data = enc;
+        el.time = yarp::os::Time::now();
+        encv = encvel_estimator->estimate(el);
+
+        //compute the orientation.
+        m_odometryData.odom_theta = (geom_r / geom_L) * (-enc[0] + enc[1]);
+
+        iCub::ctrl::AWPolyElement el2;
+        el2.data = yarp::sig::Vector(1, m_odometryData.odom_theta);
+        el2.time = yarp::os::Time::now();
+        yarp::sig::Vector vvv;
+        vvv.resize(1, 0.0);
+        vvv = encw_estimator->estimate(el2);
+
+        //build the kinematics matrix
+        /*yarp::sig::Matrix kin;
+        kin.resize(3,2);
+        kin.zero();
+        kin(0, 0) = cos(odom_theta) / 2;
+        kin(0, 1) = cos(odom_theta) / 2;
+        kin(0, 2) = sin(odom_theta) / 2;
+        kin(1, 0) = sin(odom_theta) / 2;
+        kin(1, 1) = 1 / (2 * geom_L);
+        kin(1, 2) = 1 / (2 * geom_L);
+
+        yarp::sig::Vector odom_cart_vels;  //velocities expressed in the world reference frame
+        yarp::sig::Vector base_cart_vels; //velocities expressed in the base reference frame
+        odom_cart_vels  = kin*encv;
+        base_cart_vels = kin*encv; //@@@
+
+        base_vel_x     = base_cart_vels[1];
+        base_vel_y     = base_cart_vels[0];
+        base_vel_lin   = sqrt(base_vel_x*base_vel_x + base_vel_y*base_vel_y);
+        base_vel_theta = base_cart_vels[2];
+
+        odom_vel_x      = odom_cart_vels[1];
+        odom_vel_y      = odom_cart_vels[0];
+        odom_vel_theta  = odom_cart_vels[2];
+        */
```

Comment:
is this comment useful? it was there in the previous version, but if is not useful I can delete it